x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsPhysicalOperationProviders.java

```diff
@@ -142,9 +160,33 @@ private BlockLoader getBlockLoaderFor(
         return blockLoader;
     }
 
-    private MultiTypeEsField findUnionTypes(Attribute attr) {
-        if (attr instanceof FieldAttribute fa && fa.field() instanceof MultiTypeEsField multiTypeEsField) {
-            return multiTypeEsField;
+    /// A hack to pretend an unmapped field still exists.
+    private static class DefaultShardContextForUnmappedField extends DefaultShardContext {
+        private final UnmappedEsField unmappedEsField;
+
+        DefaultShardContextForUnmappedField(DefaultShardContext ctx, UnmappedEsField unmappedEsField) {
+            super(ctx.index, ctx.ctx, ctx.aliasFilter);
+            this.unmappedEsField = unmappedEsField;
+        }
+
+        @Override
+        protected MappedFieldType fieldType(String name) {
+            var superResult = super.fieldType(name);
+            return superResult == null && name.equals(unmappedEsField.getName())
+                ? new KeywordFieldMapper.KeywordFieldType(name, false /* isIndexed */, false /* hasDocValues */, Map.of() /* meta */)
+                : superResult;
+        }
+    }
+
+    private static @Nullable MultiTypeEsField findUnionTypes(Attribute attr) {
+        if (attr instanceof FieldAttribute fa) {
+            EsField field = fa.field();
+            if (Objects.requireNonNull(field) instanceof UnmappedEsField unmapped
+                && unmapped.getState() instanceof UnmappedEsField.MultiType mt) {
+                return mt.multiTypeEsField();
```

Comment:
I see this support for combined unmapped and union-types, but did not see csv-spec or logical planner tests for this case. Is it only partially implemented at this point?